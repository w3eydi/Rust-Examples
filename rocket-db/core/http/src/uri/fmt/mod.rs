//! Type safe and URI safe formatting types and traits.

mod uri_display;
mod formatter;
mod from_uri_param;
mod encoding;
mod part;

pub use self::formatter::*;
pub use self::uri_display::*;
pub use self::from_uri_param::*;
pub use self::part::*;

pub(crate) use self::encoding::*;
