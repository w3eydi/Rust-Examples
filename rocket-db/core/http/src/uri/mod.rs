//! Types for URIs and traits for rendering URI components.

#[macro_use]
mod uri;
mod origin;
mod reference;
mod authority;
mod absolute;
mod segments;
mod path_query;
mod asterisk;
mod host;

pub mod error;
pub mod fmt;

#[doc(inline)]
pub use self::error::Error;

pub use self::uri::*;
pub use self::authority::*;
pub use self::origin::*;
pub use self::absolute::*;
pub use self::segments::*;
pub use self::reference::*;
pub use self::path_query::*;
pub use self::asterisk::*;
pub use self::host::*;
