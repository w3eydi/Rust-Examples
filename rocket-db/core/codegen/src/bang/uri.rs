use std::fmt::Display;

use devise::{Result, ext::{SpanDiagnosticExt, quote_respanned}};
use syn::{Expr, Ident, Type, spanned::Spanned};
use proc_macro2::TokenStream;

use crate::http::uri::fmt;
use crate::http_codegen::Optional;
use crate::syn_ext::IdentExt;
use crate::bang::uri_parsing::*;
use crate::attribute::param::Parameter;
use crate::exports::*;
use crate::URI_MACRO_PREFIX;

macro_rules! p {
    (@go $num:expr, $singular:expr, $plural:expr) => (
        if $num == 1 { $singular.into() } else { $plural }
    );

    ("parameter", $n:expr) => (p!(@go $n, "parameter", "parameters"));
    ($n:expr, "was") => (p!(@go $n, "1 was", format!("{} were", $n)));
    ($n:expr, "parameter") => (p!(@go $n, "1 parameter", format!("{} parameters", $n)));
}

pub fn prefix_last_segment(path: &mut syn::Path, prefix: &str) {
    let mut last_seg = path.segments.last_mut().expect("syn::Path has segments");
    last_seg.ident = last_seg.ident.prepend(prefix);
}

pub fn _uri_macro(input: TokenStream) -> Result<TokenStream> {
    let input2: TokenStream = input.clone();
    match syn::parse2::<UriMacro>(input)? {
        UriMacro::Routed(ref mut mac) => {
            prefix_last_segment(&mut mac.route.path, URI_MACRO_PREFIX);
            let path = &mac.route.path;
            Ok(quote!(#path!(#input2)))
        },
        UriMacro::Literal(uri) => Ok(quote!(#uri)),
    }
}

fn extract_exprs(internal: &InternalUriParams) -> Result<(
        impl Iterator<Item = &Expr>,             // path exprs
        impl Iterator<Item = &ArgExpr>,          // query exprs
        impl Iterator<Item = (&Ident, &Type)>,   // types for both path || query
    )>
{
    let route_name = &internal.uri_mac.route.path;
    match internal.validate() {
        Validation::Ok(exprs) => {
            let path_params = internal.dynamic_path_params();
            let path_param_count = path_params.clone().count();
            for expr in exprs.iter().take(path_param_count) {
                if expr.as_expr().is_none() {
                    return Err(expr.span().error("path parameters cannot be ignored"));
                }
            }

            let query_exprs = exprs.clone().into_iter().skip(path_param_count);
            let path_exprs = exprs.into_iter().map(|e| e.unwrap_expr()).take(path_param_count);
            let types = internal.fn_args.iter().map(|a| (&a.ident, &a.ty));
            Ok((path_exprs, query_exprs, types))
        }
        Validation::NamedIgnored(_) => {
            let mut route_name = quote!(#route_name).to_string();
            route_name.retain(|c| !c.is_whitespace());

            let diag = internal.uri_mac.args_span()
                .error("expected unnamed arguments due to ignored parameters")
                .note(format!("uri for route `{}` ignores path parameters: \"{}\"",
                        route_name, internal.route_uri));

            Err(diag)
        }
        Validation::Unnamed(expected, actual) => {
            let mut route_name = quote!(#route_name).to_string();
            route_name.retain(|c| !c.is_whitespace());

            let diag = internal.uri_mac.args_span()
                .error(format!("route expects {} but {} supplied",
                         p!(expected, "parameter"), p!(actual, "was")))
                .note(format!("route `{}` has uri \"{}\"", route_name, internal.route_uri));

            Err(diag)
        }
        Validation::Named(missing, extra, dup) => {
            let e = format!("invalid parameters for `{}` route uri", quote!(#route_name));
            let mut diag = internal.uri_mac.args_span().error(e)
                .note(format!("uri parameters are: {}", internal.fn_args_str()));

            fn join<S: Display, T: Iterator<Item = S>>(iter: T) -> (&'static str, String) {
                let mut items: Vec<_> = iter.map(|i| format!("`{}`", i)).collect();
                items.dedup();
                (p!("parameter", items.len()), items.join(", "))
            }

            if !missing.is_empty() {
                let (ps, msg) = join(missing.iter());
                diag = diag.help(format!("missing {}: {}", ps, msg));
            }

            if !extra.is_empty() {
                let (ps, msg) = join(extra.iter());
                let spans: Vec<_> = extra.iter().map(|ident| ident.span()).collect();
                diag = diag.span_help(spans, format!("unknown {}: {}", ps, msg));
            }

            if !dup.is_empty() {
                let (ps, msg) = join(dup.iter());
                let spans: Vec<_> = dup.iter().map(|ident| ident.span()).collect();
                diag = diag.span_help(spans, format!("duplicate {}: {}", ps, msg));
            }

            Err(diag)
        }
    }
}

fn add_binding<P: fmt::Part>(to: &mut Vec<TokenStream>, ident: &Ident, ty: &Type, expr: &Expr) {
    let span = expr.span();
    let part = match P::KIND {
        fmt::Kind::Path => quote_spanned!(span => #_fmt::Path),
        fmt::Kind::Query  => quote_spanned!(span => #_fmt::Query),
    };

    let tmp_ident = ident.clone().with_span(expr.span());
    let let_stmt = quote_spanned!(span => let #tmp_ident = #expr);

    to.push(quote_spanned!(span =>
        #[allow(non_snake_case)] #let_stmt;
        let #ident = <#ty as #_fmt::FromUriParam<#part, _>>::from_uri_param(#tmp_ident);
    ));
}

fn explode_path<'a>(
    internal: &InternalUriParams,
    bindings: &mut Vec<TokenStream>,
    mut exprs: impl Iterator<Item = &'a Expr>,
    mut args: impl Iterator<Item = (&'a Ident, &'a Type)>,
) -> TokenStream {
    if internal.dynamic_path_params().count() == 0 {
        let path = internal.route_uri.path().as_str();
        quote!(#_fmt::UriArgumentsKind::Static(#path))
    } else {
        let uri_display = quote!(#_fmt::UriDisplay<#_fmt::Path>);
        let dyn_exprs = internal.path_params.iter().map(|param| {
            match param {
                Parameter::Static(name) => {
                    quote!(&#name as &dyn #uri_display)
                },
                Parameter::Dynamic(_) | Parameter::Guard(_) => {
                    let (ident, ty) = args.next().expect("ident/ty for non-ignored");
                    let expr = exprs.next().expect("one expr per dynamic arg");
                    add_binding::<fmt::Path>(bindings, &ident, &ty, &expr);
                    quote_spanned!(expr.span() => &#ident as &dyn #uri_display)
                }
                Parameter::Ignored(_) => {
                    let expr = exprs.next().expect("one expr per dynamic arg");
                    quote_spanned!(expr.span() => &#expr as &dyn #uri_display)
                }
            }
        });

        quote!(#_fmt::UriArgumentsKind::Dynamic(&[#(#dyn_exprs),*]))
    }
}

fn explode_query<'a>(
    internal: &InternalUriParams,
    bindings: &mut Vec<TokenStream>,
    mut arg_exprs: impl Iterator<Item = &'a ArgExpr>,
    mut args: impl Iterator<Item = (&'a Ident, &'a Type)>,
) -> Option<TokenStream> {
    let query = internal.route_uri.query()?.as_str();
    if internal.dynamic_query_params().count() == 0 {
        return Some(quote!(#_fmt::UriArgumentsKind::Static(#query)));
    }

    let query_arg = quote!(#_fmt::UriQueryArgument);
    let uri_display = quote!(#_fmt::UriDisplay<#_fmt::Query>);
    let dyn_exprs = internal.query_params.iter().filter_map(|param| {
        if let Parameter::Static(source) = param {
            return Some(quote!(#query_arg::Raw(#source)));
        }

        let dynamic = match param {
            Parameter::Static(source) =>  {
                return Some(quote!(#query_arg::Raw(#source)));
            },
            Parameter::Dynamic(ref seg) => seg,
            Parameter::Guard(ref seg) => seg,
            Parameter::Ignored(_) => unreachable!("invariant: unignorable q")
        };

        let (ident, ty) = args.next().expect("ident/ty for query");
        let arg_expr = arg_exprs.next().expect("one expr per query");
        let expr = match arg_expr.as_expr() {
            Some(expr) => expr,
            None => {
                // Force a typecheck for the `Ignoreable` trait.
                bindings.push(quote_respanned! { arg_expr.span() =>
                    #_fmt::assert_ignorable::<#_fmt::Query, #ty>();
                });

                return None;
            }
        };

        let name = &dynamic.name;
        add_binding::<fmt::Query>(bindings, &ident, &ty, &expr);
        Some(match dynamic.trailing {
            false => quote_spanned! { expr.span() =>
                #query_arg::NameValue(#name, &#ident as &dyn #uri_display)
            },
            true => quote_spanned! { expr.span() =>
                #query_arg::Value(&#ident as &dyn #uri_display)
            },
        })
    });

    Some(quote!(#_fmt::UriArgumentsKind::Dynamic(&[#(#dyn_exprs),*])))
}

pub fn _uri_internal_macro(input: TokenStream) -> Result<TokenStream> {
    // Parse the internal invocation and the user's URI param expressions.
    let internal = syn::parse2::<InternalUriParams>(input)?;
    let (path_exprs, query_exprs, mut fn_args) = extract_exprs(&internal)?;

    let mut bindings = vec![];
    let path = explode_path(&internal, &mut bindings, path_exprs, &mut fn_args);
    let query = Optional(explode_query(&internal, &mut bindings, query_exprs, fn_args));

    let prefix = internal.uri_mac.prefix.as_ref()
        .map(|prefix| quote_spanned!(prefix.span() => .with_prefix(#prefix)));

    let suffix = internal.uri_mac.suffix.as_ref()
        .map(|suffix| quote_spanned!(suffix.span() => .with_suffix(#suffix)));

     Ok(quote_spanned!(internal.uri_mac.route.path.span() =>
         #[allow(unused_braces)] {
             #(#bindings)*
             let __builder = #_fmt::RouteUriBuilder::new(#path, #query);
             __builder #prefix #suffix .render()
         }
     ))
}
