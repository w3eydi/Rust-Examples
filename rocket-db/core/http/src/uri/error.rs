//! Errors arising from parsing invalid URIs.

use std::fmt;

pub use crate::parse::uri::Error;

/// The error type returned when a URI conversion fails.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct TryFromUriError(pub(crate) ());

impl fmt::Display for TryFromUriError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        "invalid conversion from general to specific URI variant".fmt(f)
    }
}

/// An error interpreting a segment as a [`PathBuf`] component in
/// [`Segments::to_path_buf()`].
///
/// [`PathBuf`]: std::path::PathBuf
/// [`Segments::to_path_buf()`]: crate::uri::Segments::to_path_buf()
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum PathError {
    /// The segment started with the wrapped invalid character.
    BadStart(char),
    /// The segment contained the wrapped invalid character.
    BadChar(char),
    /// The segment ended with the wrapped invalid character.
    BadEnd(char),
}
