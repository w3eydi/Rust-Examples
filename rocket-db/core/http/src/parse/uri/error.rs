use std::{fmt, convert};
use std::borrow::Cow;

use pear::error::Expected;
use pear::input::ParseError;
use crate::parse::uri::RawInput;
use crate::ext::IntoOwned;

/// Error emitted on URI parse failure.
///
/// Internally, the type includes information about where the parse error
/// occurred (the error's context) and information about what went wrong.
/// Externally, this information can be retrieved (in textual form) through its
/// `Display` implementation. In other words, by printing a value of this type.
#[derive(Debug)]
pub struct Error<'a> {
    pub(crate) expected: Expected<u8, Cow<'a, [u8]>>,
    pub(crate) index: usize,
}

#[doc(hidden)]
impl<'a> From<ParseError<RawInput<'a>>> for Error<'a> {
    fn from(inner: ParseError<RawInput<'a>>) -> Self {
        let expected = inner.error.map(convert::identity, |v| v.values.into());
        Error { expected, index: inner.info.context.start }
    }
}

impl Error<'_> {
    /// Returns the byte index into the text where the error occurred if it is
    /// known.
    ///
    /// # Example
    ///
    /// ```rust
    /// # extern crate rocket;
    /// use rocket::http::uri::Origin;
    ///
    /// let err = Origin::parse("/foo bar").unwrap_err();
    /// assert_eq!(err.index(), 4);
    /// ```
    pub fn index(&self) -> usize {
        self.index
    }
}

impl fmt::Display for Error<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} at index {}", self.expected, self.index)
    }
}

impl IntoOwned for Error<'_> {
    type Owned = Error<'static>;

    fn into_owned(self) -> Error<'static> {
        Error {
            expected: self.expected.map(|t| t, |s| s.into_owned().into()),
            index: self.index
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::parse::uri::origin_from_str;

    macro_rules! check_err {
        ($url:expr => $error:expr) => {{
            let e = origin_from_str($url).unwrap_err();
            assert_eq!(e.to_string(), $error.to_string())
        }}
    }

    #[test]
    fn check_display() {
        check_err!("a" => "expected token '/' but found 'a' at index 0");
        check_err!("?" => "expected token '/' but found '?' at index 0");
        check_err!("è¿™" => "expected token '/' but found byte 232 at index 0");
    }
}
