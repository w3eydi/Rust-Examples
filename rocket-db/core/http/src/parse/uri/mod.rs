mod parser;
mod error;
pub(crate) mod tables;

#[cfg(test)] mod tests;

use crate::uri::{Uri, Origin, Absolute, Authority, Reference, Asterisk};

use self::parser::*;

pub use self::error::Error;

type RawInput<'a> = pear::input::Pear<pear::input::Cursor<&'a [u8]>>;

#[inline]
pub fn from_str(s: &str) -> Result<Uri<'_>, Error<'_>> {
    Ok(parse!(uri: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn origin_from_str(s: &str) -> Result<Origin<'_>, Error<'_>> {
    Ok(parse!(origin: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn authority_from_str(s: &str) -> Result<Authority<'_>, Error<'_>> {
    Ok(parse!(authority: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn authority_from_bytes(s: &[u8]) -> Result<Authority<'_>, Error<'_>> {
    Ok(parse!(authority: RawInput::new(s))?)
}

#[inline]
pub fn scheme_from_str(s: &str) -> Result<&str, Error<'_>> {
    let _validated = parse!(scheme: RawInput::new(s.as_bytes()))?;
    Ok(s)
}

#[inline]
pub fn absolute_from_str(s: &str) -> Result<Absolute<'_>, Error<'_>> {
    Ok(parse!(absolute: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn asterisk_from_str(s: &str) -> Result<Asterisk, Error<'_>> {
    Ok(parse!(asterisk: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn reference_from_str(s: &str) -> Result<Reference<'_>, Error<'_>> {
    Ok(parse!(reference: RawInput::new(s.as_bytes()))?)
}
