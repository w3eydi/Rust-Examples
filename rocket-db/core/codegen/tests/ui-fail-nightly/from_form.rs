../ui-fail/from_form.rs