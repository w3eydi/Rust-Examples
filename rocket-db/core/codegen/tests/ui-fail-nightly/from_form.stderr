error: enums are not supported
 --> $DIR/from_form.rs:4:1
  |
4 | enum Thing { }
  | ^^^^^^^^^^^^^^
  |
note: error occurred while deriving `FromForm`
 --> $DIR/from_form.rs:3:10
  |
3 | #[derive(FromForm)]
  |          ^^^^^^^^
  = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: at least one field is required
 --> $DIR/from_form.rs:7:1
  |
7 | struct Foo1;
  | ^^^^^^^^^^^^
  |
note: error occurred while deriving `FromForm`
 --> $DIR/from_form.rs:6:10
  |
6 | #[derive(FromForm)]
  |          ^^^^^^^^
  = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: at least one field is required
  --> $DIR/from_form.rs:10:13
   |
10 | struct Foo2 {  }
   |             ^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:9:10
   |
9  | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: tuple struct must have exactly one field
  --> $DIR/from_form.rs:16:12
   |
16 | struct Foo4(usize, usize, usize);
   |            ^^^^^^^^^^^^^^^^^^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:15:10
   |
15 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only one lifetime is supported
  --> $DIR/from_form.rs:19:25
   |
19 | struct NextTodoTask<'f, 'a> {
   |                         ^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:18:10
   |
18 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
  --> $DIR/from_form.rs:28:20
   |
28 |     #[field(name = "isindex")]
   |                    ^^^^^^^^^
   |
   = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:26:10
   |
26 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field name conflicts with previous name
  --> $DIR/from_form.rs:36:5
   |
36 |     foo: usize,
   |     ^^^
   |
help: declared in this field
  --> $DIR/from_form.rs:36:5
   |
36 |     foo: usize,
   |     ^^^^^^^^^^
note: previous field with conflicting name
  --> $DIR/from_form.rs:34:5
   |
34 | /     #[field(name = "foo")]
35 | |     field: String,
   | |_________________^
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:32:10
   |
32 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field name conflicts with previous name
  --> $DIR/from_form.rs:43:20
   |
43 |     #[field(name = "hello")]
   |                    ^^^^^^^
   |
help: declared in this field
  --> $DIR/from_form.rs:43:5
   |
43 | /     #[field(name = "hello")]
44 | |     other: String,
   | |_________________^
note: previous field with conflicting name
  --> $DIR/from_form.rs:41:5
   |
41 | /     #[field(name = "hello")]
42 | |     first: String,
   | |_________________^
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:39:10
   |
39 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field name conflicts with previous name
  --> $DIR/from_form.rs:50:20
   |
50 |     #[field(name = "first")]
   |                    ^^^^^^^
   |
help: declared in this field
  --> $DIR/from_form.rs:50:5
   |
50 | /     #[field(name = "first")]
51 | |     other: String,
   | |_________________^
note: previous field with conflicting name
  --> $DIR/from_form.rs:49:5
   |
49 |     first: String,
   |     ^^^^^^^^^^^^^
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:47:10
   |
47 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected attribute parameter: `field`
  --> $DIR/from_form.rs:56:28
   |
56 |     #[field(name = "blah", field = "bloo")]
   |                            ^^^^^^^^^^^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:54:10
   |
54 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected list `#[field(..)]`, found bare path "field"
  --> $DIR/from_form.rs:62:7
   |
62 |     #[field]
   |       ^^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:60:10
   |
60 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected key/value `key = value`
  --> $DIR/from_form.rs:68:13
   |
68 |     #[field("blah")]
   |             ^^^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:66:10
   |
66 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected key/value `key = value`
  --> $DIR/from_form.rs:74:13
   |
74 |     #[field(123)]
   |             ^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:72:10
   |
72 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected attribute parameter: `beep`
  --> $DIR/from_form.rs:80:13
   |
80 |     #[field(beep = "bop")]
   |             ^^^^^^^^^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:78:10
   |
78 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field has conflicting names
  --> $DIR/from_form.rs:86:5
   |
86 | /     #[field(name = "blah")]
87 | |     #[field(name = "blah")]
88 | |     my_field: String,
   | |____________________^
   |
note: this field name...
  --> $DIR/from_form.rs:86:20
   |
86 |     #[field(name = "blah")]
   |                    ^^^^^^
note: ...conflicts with this field name
  --> $DIR/from_form.rs:87:20
   |
87 |     #[field(name = "blah")]
   |                    ^^^^^^
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:84:10
   |
84 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected list `#[attr(..)]`, found bare boolean literal
  --> $DIR/from_form.rs:93:20
   |
93 |     #[field(name = true)]
   |                    ^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:91:10
   |
91 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected expression, found bare path "name"
  --> $DIR/from_form.rs:99:13
   |
99 |     #[field(name)]
   |             ^^^^
   |
note: error occurred while deriving `FromForm`
  --> $DIR/from_form.rs:97:10
   |
97 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected list `#[attr(..)]`, found bare integer literal
   --> $DIR/from_form.rs:105:20
    |
105 |     #[field(name = 123)]
    |                    ^^^
    |
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:103:10
    |
103 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
   --> $DIR/from_form.rs:111:20
    |
111 |     #[field(name = "hello&world")]
    |                    ^^^^^^^^^^^^^
    |
    = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:109:10
    |
109 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
   --> $DIR/from_form.rs:117:20
    |
117 |     #[field(name = "!@#$%^&*()_")]
    |                    ^^^^^^^^^^^^^
    |
    = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:115:10
    |
115 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
   --> $DIR/from_form.rs:123:20
    |
123 |     #[field(name = "?")]
    |                    ^^^
    |
    = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:121:10
    |
121 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
   --> $DIR/from_form.rs:129:20
    |
129 |     #[field(name = "")]
    |                    ^^
    |
    = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:127:10
    |
127 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
   --> $DIR/from_form.rs:135:20
    |
135 |     #[field(name = "a&b")]
    |                    ^^^^^
    |
    = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:133:10
    |
133 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid form field name
   --> $DIR/from_form.rs:141:20
    |
141 |     #[field(name = "a=")]
    |                    ^^^^
    |
    = help: field name cannot be `isindex` or contain '&', '=', '?', '.', '[', ']'
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:139:10
    |
139 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: duplicate attribute parameter: default
   --> $DIR/from_form.rs:177:26
    |
177 |     #[field(default = 1, default = 2)]
    |                          ^^^^^^^^^^^
    |
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:175:10
    |
175 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: duplicate default field expression
   --> $DIR/from_form.rs:184:23
    |
184 |     #[field(default = 2)]
    |                       ^
    |
    = help: at most one `default` or `default_with` is allowed
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:181:10
    |
181 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: duplicate default expressions
   --> $DIR/from_form.rs:190:23
    |
190 |     #[field(default = 1, default_with = None)]
    |                       ^
    |
    = help: only one of `default` or `default_with` must be used
note: other default expression is here
   --> $DIR/from_form.rs:190:41
    |
190 |     #[field(default = 1, default_with = None)]
    |                                         ^^^^
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:188:10
    |
188 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: duplicate default expressions
   --> $DIR/from_form.rs:197:23
    |
197 |     #[field(default = 1)]
    |                       ^
    |
    = help: only one of `default` or `default_with` must be used
note: other default expression is here
   --> $DIR/from_form.rs:196:28
    |
196 |     #[field(default_with = None)]
    |                            ^^^^
note: error occurred while deriving `FromForm`
   --> $DIR/from_form.rs:194:10
    |
194 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find function `unknown` in this scope
   --> $DIR/from_form.rs:153:24
    |
153 |     #[field(validate = unknown())]
    |                        ^^^^^^^ not found in this scope

error[E0308]: mismatched types
   --> $DIR/from_form.rs:147:24
    |
147 |     #[field(validate = 123)]
    |             --------   ^^^ expected enum `Result`, found integer
    |             |
    |             expected due to this
    |
    = note: expected enum `Result<(), Errors<'_>>`
               found type `{integer}`

error[E0308]: mismatched types
   --> $DIR/from_form.rs:160:12
    |
160 |     first: String,
    |            ^^^^^^ expected enum `TempFile`, found struct `std::string::String`
    |
    = note: expected reference `&TempFile<'_>`
               found reference `&std::string::String`

error[E0308]: mismatched types
   --> $DIR/from_form.rs:166:12
    |
166 |     first: String,
    |            ^^^^^^ expected enum `TempFile`, found struct `std::string::String`
    |
    = note: expected reference `&TempFile<'_>`
               found reference `&std::string::String`

error[E0308]: mismatched types
   --> $DIR/from_form.rs:165:28
    |
165 |     #[field(validate = ext("hello"))]
    |                            ^^^^^^^ expected struct `ContentType`, found `&str`

error[E0277]: the trait bound `i32: From<&str>` is not satisfied
   --> $DIR/from_form.rs:171:23
    |
171 |     #[field(default = "no conversion")]
    |                       ^^^^^^^^^^^^^^^ the trait `From<&str>` is not implemented for `i32`
    |
    = help: the following implementations were found:
              <i32 as From<NonZeroI32>>
              <i32 as From<bool>>
              <i32 as From<i16>>
              <i32 as From<i8>>
            and 5 others
    = note: required because of the requirements on the impl of `Into<i32>` for `&str`

error[E0308]: mismatched types
   --> $DIR/from_form.rs:203:33
    |
203 |     #[field(default_with = Some("hi"))]
    |                                 ^^^^ expected struct `std::string::String`, found `&str`
    |
help: try using a conversion method
    |
203 |     #[field(default_with = Some("hi".to_string()))]
    |                                     ++++++++++++
203 |     #[field(default_with = Some("hi".to_string()))]
    |                                     ++++++++++++
